docs/manpages/NekMesh.1
.\" Manpage for ADRSolver
.\" Contact c.cantwell@imperial.ac.uk to correct errors or typos
.TH man 1 "07 Aug 2019" "5.0" "NekMesh man page"
.SH NAME
NekMesh \- generates, converts and manipulates high-order meshes
.SH SYNOPSIS
.B NekMesh
[option(s)] input output
.SH DESCRIPTION
.B NekMesh
is a high-order finite element mesh generator and general pre-processing
mesh utility, that can be used to either generate meshes directly from an
underlying CAD representation of a geometry, or alternatively to import and
process linear meshes from a range of other mesh generation tools to make them
suitable for high-order simulations.

.B NekMesh
is designed around the concept of modules that run sequentially in a
pipeline.
.B Input modules
read either a file from an external mesh generation utility, or a
.B NekMesh
.B .mcf
mesh configuration file which can be used to read CAD geometry. The input module
is selected based on the input filename.
Optionally, one or more
.B processing modules
can be used to modify the mesh in some way, for e.g. the creation of boundary
layers or the optimisation of the input mesh. Finally, a
.B output module
is used to write out the resulting mesh, depending on the extension of the
filename. All modules typically support a number of options that can be defined
on the command line.

.SH OPTIONS
.TP 4
.B \-h [ \-\-help ]
Prints a help message and program synopsys.
.TP 4
.B \-l [ \-\-modules-list ]
Prints a list of all input, processing and output modules that can be used
within NekMesh.
.TP 4
.B \-p [ \-\-modules-opt ] module
Prints the list of options for a given module.
.TP 4
.B \-m [ \-\-module ] module
Specifies a module to be used. Modules are run in the order that they are
defined in the command line options.
.TP 4
.B \-v [ \-\-verbose ]
Print additional debugging and mesh generation information to screen.
.SH BUGS
No known bugs.
.SH EXAMPLE
As an example, in the simplest case
.B NekMesh
can be used to simply convert files between one format and another; for example
reading a Gmsh file and writing a Nektar++ format file can be achieved with the
command

.RS
.B NekMesh input.msh output.xml
.RE

As a more complex example, the command below:

.RS
.B NekMesh -m varopti:hyperelastic:nq=4 -m jac:list input.gmsh:msh output.xml:xml:uncompress
.RE

runs the following pipeline:

.RS
.IP \(em 3
Load the input.gmsh file as a Gmsh module;

.IP \(em 3
Run the varopti module with the boolean option 'hyperelastic' and setting the
quadrature order to 4;

.IP \(em 3
List any elements that have negative Jacobian with the jac module;

.IP \(em 3
Finally output the mesh as a Nektar++ XML file in an uncompressed form.
.RE

For a list of modules and their options, the
.B \-l
and
.B \-p
options can be used respectively. More advanced usage of
.B NekMesh
can be found in the accompanying user guide.

.SH AUTHOR
David Moxey (d.moxey@exeter.ac.uk)


CHANGELOG.md
Changelog
=========

v5.1.0
------
**Library**
- Refactored time integration code using factory pattern (!1034)
- Fix to preprocessor logic for boost with Visual Studio >= 2015 (!1115)

**CardiacEPSolver**
- Added additional parameter sets to Fenton-Karma model (!1119)

**NekMesh**
- Improved boundary layer splitting and output to CADfix (!938)

v5.0.1
------
**Library**
- Fix incorrect coordinate dimension used in history point filter (!1118)
v5.0.1
------
- Fix compile errors with GCC 9.x (!1108)

v5.0.1
------
**CI**
- Switched to Gitlab CI (!1120, !1120, !1128, !1129)

v5.0.0
------
**Library**
- Added in sum factorisation version for pyramid expansions and orthogonal
  expansion in pyramids (!750)
- Added detection of 'abort' file to cleanly terminate simulation early (!772)
- Significant overhaul of CMake infrastructure (!770, !804)
- Fix ThridpartyCCM options (!802)
- Fix Windows CRLF tokens in GEO reader and improve comment handling (!805)
- Use chrono in Timer (!807)
- Fix caching of FUNCTION tags that read from file and provide the same
  functionality in FUNCTIONs defined for forcings (!759)
- Transition to C++11 (!795, !847)
- Add patch to tinyxml to fix size_t vs int bug (!820, !1006)
- Add ARPACK thirdparty build capabilities (!828)
- Added native support for csv files in addititon to pts (!760, !835, !906)
- Utilize LAPACK_DIR env variable to find the native blas/lapack install (!827)
- Extend AeroForces filter to compressible flows (!815)
- Remove StdExpansion use from MultiRegion (use Expansions instead). (!831)
- Move steady state check and CFL output from solvers to SolverUtils (!832)
- Remove DG advection implementation from EquationSystem (!832)
- Simplify RawType typedefs (!840)
- Remove unused files from BasicUtils (!841)
- Remove checks for old boost versions which are no longer supported (!841)
- Refactor ParseUtils to be more consistent (!843, !896, !908)
- Added support for using the distance to a specific region (e.g. outlet) in the
  function definitions for the Absorption Forcing (!769)
- Improve performance of DisContField2D::v_ExtractTracePhys (!824)
- Fix small bug in Jacobian Energy (!857)
- fix variable name overriding in file functions (!870)
- Adds CFI CAD engine back-end (!864)
- Adds CFI Mesh IO support (!864)
- Cleanup of CAD system data structures (!864)
- Fix mac OSX on buildbots (!876)
- Fix error from (!826) (!876)
- Fix minor bug in ARPACK thirdparty build cmake (!874)
- Added in sum factorisation version for pyramid expnasions and orthogonal
  expansion in pyramids (!750)
- Adjust boost third-party compilation to account for different toolset
  choices (!886)
- Switch MeshGraph to use factory pattern and add HDF5 geometry support (!900,
  !904, !941)
- Restructure the low energy preconditioner to handle pyramidic and variable
  p expansions (!920)
- Remove requirement for modmetis, switch to SCOTCH by default (!899)
- Switch MeshGraph to use factory pattern and add HDF5 geometry support
  (!900, !904)
- Fix bug in MeshPartition.cpp which caused incorrect array access when
  WeightPartitions was used in parallel (!923)
- Removed instance count from beginning of Array storage to improve memory
  alignment (!921)
- Fix naming issue of duplicate Unit tests (!924)
- Fix warnings about missing virtual destructors in abstract classes (!932)
- Fix ability to have periodic boundary conditions that are aligned by a
  rotation rather than just a translation (!933)
- Added a coupling interface to exchange data between solvers at run time
  and a DummySolver to test the implementations (!853, !931, !950, !973, !1017)
- Fix compilation issue with newer Boost versions and clang (!940)
- If only `NEKTAR_BUILD_LIBRARY` is enabled, only libraries up to and including
  `MultiRegions` will be built by default (!945)
- Dont add doxygen documentation to the all target (!834)
- Fix missing metadata import from Hdf5 files (!971)
- Fix missing flags for periodic BC in DiffusionLDG (!985)
- Add the moving reference frame as a forcing (!987)
- Added rtree for element bounding box lookup to accelerate interpolation (!996,
  !1066)
- Fix integration weights on prisms and pyramids if not using the default
  integration rule (!998)
- Fix missing ContainsPoint in Pyramid expansion (!1000)
- Added path prefixes to find packaged Scotch (!979, !1008)
- Add HDF5 geometry format (!977)
- Combine and generalise demo code in StdRegions and LocalRegions (!993)
- Fix for error output to allow for custom error streams (!944)
- Fixed bug in ReOrientQuadFacePhysMap (!1003)
- Add NekPy Python interface (!962, !990, !989, !1004, !1014, !1061, !1070)
- Fix edge case for ThirdPartyScotch and FindScoth (!1009)
- Fix to populate m_elmtToExpId map if not already set up in GetExpIndex (!1019)
- Added flag to skip periodic BCs while filling Dirichlet BCs in
  ContField3D.cpp (!1018)
- Fix bounding box for interpolation (!1033)
- Added IMEXOrder4, RK5 and AB4 time integration schemes (!1037)
- Fix TriExp.cpp orientation bug (!1048)
- Fix XML attributes in conditions.cpp to be unordered (!1015)
- Fix issue with HDF5 mesh input in serial (!1049)
- Add estimate of filters CPU time (!1044)
- Update CompressibleFlowSolver/Examples/Test_IsentropicVortex1.xml example (!1045)
- Add error if HDG used with periodic BCs (!1071)
- Fix issues related to leading factors, arithmetic order and associativity of
  exponential operator in expression evaluator (!1066)
- Remove use of `using namespace std` in header files (!1066)
- Add error messages for use of ARPACK in serial (!1079)
- Generalise ContainsPoint routine (!1078)
- Homogenized fallthrough to fix issues with gcc 7.4.0 (!1084)

**NekMesh**:
- Add feature to read basic 2D geo files as CAD (!731)
- Add periodic boundary condition meshing in 2D (!733)
- Adjust boundary layer thickness in corners in 2D (!739)
- Add non-O BL meshing in 2D (!757)
- Add ability to compile CCIO library but tar file is not yet openly

utilities/NekMesh/NekMesh.cpp
////////////////////////////////////////////////////////////////////////////////
//
//  File: NekMesh.cpp
//
//  For more information, please see: http://www.nektar.info/
//
//  The MIT License
//
//  Copyright (c) 2006 Division of Applied Mathematics, Brown University (USA),
//  Department of Aeronautics, Imperial College London (UK), and Scientific
//  Computing and Imaging Institute, University of Utah (USA).
//
//  License for the specific language governing rights and limitations under
//  Permission is hereby granted, free of charge, to any person obtaining a
//  copy of this software and associated documentation files (the "Software"),
//  to deal in the Software without restriction, including without limitation
//  the rights to use, copy, modify, merge, publish, distribute, sublicense,
//  and/or sell copies of the Software, and to permit persons to whom the
//  Software is furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included
//  in all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
//  OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
//  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
//  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
//  DEALINGS IN THE SOFTWARE.
//
//  Description: Mesh conversion utility.
//
////////////////////////////////////////////////////////////////////////////////

#include <string>
#include <chrono>
#include <boost/algorithm/string.hpp>
#include <LibUtilities/BasicConst/GitRevision.h>
#include <LibUtilities/BasicUtils/Timer.h>
#include <boost/program_options.hpp>
#include <boost/asio/ip/host_name.hpp>
#include <boost/format.hpp>

#include <NekMeshUtils/Module/Module.h>

using namespace std;
using namespace Nektar::NekMeshUtils;

namespace po = boost::program_options;
namespace ip    = boost::asio::ip;

int main(int argc, char* argv[])
{
    po::options_description desc("Available options");
    desc.add_options()
        ("help,h",         "Produce this help message.")
        ("modules-list,l", "Print the list of available modules.")
        ("modules-opt,p",  po::value<string>(),
             "Print options for a module.")
        ("module,m",       po::value<vector<string> >(),
             "Specify modules which are to be used.")
        ("verbose,v",      "Enable verbose mode.");

    po::options_description hidden("Hidden options");
    hidden.add_options()
        ("input-file",   po::value<vector<string> >(), "Input filename");

    po::options_description cmdline_options;
    cmdline_options.add(hidden).add(desc);

    po::options_description visible("Allowed options");
    visible.add(desc);

    po::positional_options_description p;
    p.add("input-file", -1);

    po::variables_map vm;

    try
    {
        po::store(po::command_line_parser(argc, argv).
                  options(cmdline_options).positional(p).run(), vm);
        po::notify(vm);
    }
    catch (const std::exception& e)
    {
        cerr << e.what() << endl;
        cerr << desc;
        return 1;
    }

    // Print available modules.
    if (vm.count("modules-list"))
    {
        GetModuleFactory().PrintAvailableClasses(std::cerr);
        return 1;
    }

    if (vm.count("modules-opt"))
    {
        vector<string> tmp1;
        boost::split(tmp1, vm["modules-opt"].as<string>(), boost::is_any_of(":"));

        if (tmp1.size() != 2)
        {
            cerr << "ERROR: To specify a module, use one of in, out or proc "
                 << "together with the filename; for example in:vtk." << endl;
            return 1;
        }

        if (tmp1[0] != "in" && tmp1[0] != "out" && tmp1[0] != "proc")
        {
            cerr << "ERROR: Invalid module type " << tmp1[0] << endl;
            return 1;
        }

        ModuleType t;

        if (tmp1[0] == "in")
        {
            t = eInputModule;
        }
        else if (tmp1[0] == "out")
        {
            t = eOutputModule;
        }
        else if (tmp1[0] == "proc")
        {
            t = eProcessModule;
        }

        MeshSharedPtr m = std::shared_ptr<Mesh>(new Mesh());
        ModuleSharedPtr mod = GetModuleFactory().CreateInstance(
            ModuleKey(t, tmp1[1]), m);
        cerr << "Options for module " << tmp1[1] << ":" << endl;
        mod->PrintConfig();
        return 1;
    }

    if (vm.count("help") || vm.count("input-file") != 1) {
        cerr << "Usage: NekMesh [options] inputfile.ext1 outputfile.ext2"
             << endl;
        cout << desc;
        return 1;
    }

    vector<string> inout = vm["input-file"].as<vector<string> >();

    if (inout.size() < 2)
    {
        cerr << "ERROR: You must specify an input and output file." << endl;
        return 1;
    }

    /*
     * Process list of modules. Each element of the vector of module strings can
     * be in the following form:
     *
     * modname:arg1=a:arg2=b:arg3=c:arg4:arg5=asd
     *
     * where the only required argument is 'modname', specifing the name of the
     * module to load.
     */

    MeshSharedPtr mesh = std::shared_ptr<Mesh>(new Mesh());

    // Add provenance information to mesh.
    stringstream ss;
    for(int i = 1; i < argc; i++)
    {
        ss << argv[i] << " ";
    }
    mesh->m_metadata["NekMeshCommandLine"] = ss.str();

    vector<ModuleSharedPtr> modules;
    vector<string>          modcmds;

    if (vm.count("verbose"))
    {
        mesh->m_verbose = true;
    }

    if (vm.count("module"))
    {
        modcmds = vm["module"].as<vector<string> >();
    }

    // Add input and output modules to beginning and end of this vector.
    modcmds.insert   (modcmds.begin(), inout[0]);
    modcmds.push_back(inout[1]);

    for (int i = 0; i < modcmds.size(); ++i)
    {
        // First split each command by the colon separator.
        vector<string> tmp1;
        ModuleKey module;
        int offset = 1;

        boost::split(tmp1, modcmds[i], boost::is_any_of(":"));

        if (i == 0 || i == modcmds.size() - 1)
        {
            module.first = (i == 0 ? eInputModule : eOutputModule);

            // If no colon detected, automatically detect mesh type from
            // file extension. Otherwise override and use tmp1[1] as the
            // module to load. This also allows us to pass options to
            // input/output modules. So, for example, to override
            // filename.xml to be read as vtk, you use:
            //
            // filename.xml:vtk:opt1=arg1:opt2=arg2
            if (tmp1.size() == 1)
            {
                int    dot    = tmp1[0].find_last_of('.') + 1;
                string ext    = tmp1[0].substr(dot, tmp1[0].length() - dot);

                if (ext == "gz")
                {
                    string tmp = tmp1[0].substr(0, tmp1[0].length() - 3);
                    dot = tmp.find_last_of('.') + 1;
                    ext = tmp.substr(dot, tmp.length() - dot);
                }

                module.second = ext;
                tmp1.push_back(string(i == 0 ? "infile=" : "outfile=")+tmp1[0]);
            }
            else
            {
                module.second = tmp1[1];
                tmp1.push_back(string(i == 0 ? "infile=" : "outfile=")+tmp1[0]);
                offset++;
            }
        }
        else
        {
            module.first  = eProcessModule;
            module.second = tmp1[0];
        }

        // Create module.
        ModuleSharedPtr mod = GetModuleFactory().CreateInstance(module,mesh);
        modules.push_back(mod);

        // Set options for this module.
        for (int j = offset; j < tmp1.size(); ++j)
        {
            vector<string> tmp2;
            boost::split(tmp2, tmp1[j], boost::is_any_of("="));

            if (tmp2.size() == 1)
            {
                mod->RegisterConfig(tmp2[0]);
            }
            else if (tmp2.size() == 2)
            {
                mod->RegisterConfig(tmp2[0], tmp2[1]);
            }
            else
            {
                cerr << "ERROR: Invalid module configuration: format is "
                     << "either :arg or :arg=val" << endl;
                abort();
            }
        }

        // Ensure configuration options have been set.
        mod->SetDefaults();
    }

    // Run mesh process.
    for (int i = 0; i < modules.size(); ++i)
    {
        Nektar::LibUtilities::Timer t;
        t.Start();
        modules[i]->Process();
        t.Stop();

        if (mesh->m_verbose)
        {
            std::cout << "Module elapsed time: " << t.TimePerTest(1) << std::endl;
        }
    }

    return 0;
}
